### ISSUE TITLE
Make it so that users can create their own vocab for the bot.

### ISSUE THREAD

**Issue by FreezePhoenix** [LunarWatcher/NN-chatbot#4]
I think it would be useful if users could make their own vocab the first time before they make saves for the bot. This would require 2 things:

- A way to access the chatroom posts.
- A way to write to a file.

The first one can be solved by SO-chatbot by Zirak, combined with the modification of the `console.log` function.
The second one can be done by opening a new window and using `document.write`.

**Comment by TheOneWhoMade** [comment]
No clue how to label this as a feature request.

**Comment by TheOneWhoMade** [comment]
this bookmark kindof shows what I'm thinking of:
`javascript:(function(window){ var a=document.createElement("SCRIPT"); a.src="https://rawgithub.com/Zirak/SO-ChatBot/master/master.js"; a.onload=function(){ bot.activateDevMode(); bot.config.pattern="$$"; bot.newwindow=window.open(); bot.users[9107868].is_owner=true; bot.users[1839506].is_owner=false; bot.users[829835].is_owner=false; bot.users[1596138].is_owner=false; }; document.head.appendChild(a);var hj=function(e){ if(e.events!=undefined){for(var i = 0;i<e.events.length;i++){try{bot.newwindow.document.write(e.events[i].content+"<br>")}catch(e){bot.newwindow.document.write(e)}}} }; console.log=hj; })(window);`

**Comment by TheOneWhoMade** [comment]
LunarWatcher do you think it would be better written in JS or Python? I think JS because it would run directly in the browser.

**Comment by LunarWatcher** [comment]
ATM, Javascript, like the JVM languages, PHP, and a bunch of other languages don't have ML libraries that come near TensorFlow. The Python port is the most extensive one. Lua would be another option (it has a library I can't remember the name of, but since I don't know Lua and TF is still the currently most comprehensive base-ML lib) Python and TF is the best option. Plus, Python and TF has other libraries (i.e. Keras, Tensorlayer (this project uses Tensorlayer), etc) that build on top of TF to add another layer of abstraction and save time by not having to create the layers. Whether or not the bot can run in the browser is of 0 interest. I could create a Flask backend for web events, but there would still be python running in the background which kinda defeats the purpose. I'm not going to re-write the code to allow bookmarklet use. Create a batch script that executes a command that runs the script. I.e.

    python <DRIVE>:/<path to>/bot.py

I can add the necessary input as CLI arguments in addition to stdin (as it currently is), but that's as far as I'm gonna go. And finally, custom vocabs are going to be supported with a currently unpushed edit, where the vocab size 45000 is used, which allows slightly above 7000 custom words in the vocab. Requires training data which I will write a scrapper for later (scrapper here refers to collecting conversations, filtering them manually and making it possible to use that to train the bot). Owners are hard-code atm, ranks are added last time I debugged it in chat, chatroom posts are sent through ChatExchange so there's no reason to have console input to send messages (unrelated: the process of creating a system that handles which site to send to is really hard to create. Since there's the site, room and message there are three vars, discord is more or less impossible due to the channel IDs being defined by long. And there's not really any benefit to manually sending messages from the account anyways, so it's not a priority.

**TL;DR:** Python will be used because JS has no good ML libraries that come close to TF. Bookmarklets aren't going to be supported ATM due to a lacking server callback (Flask) for a potential web frontend. Either way, it runs on python in the background (Flask is basically a python server) so it just adds more memory usage. Custom vocabs are prepared to be pushed atm, but I'm working on figuring out memory issues for low-memory GPU's (optimizing for 2 gigs of VRAM, but maintaining full GPU support is hard)

**Comment by TheOneWhoMade** [comment]
I don't think you understand. What I am saying is that JS could create a file that creates output similar to the original Dialog files. It would not run a bot under JS, just create a file ready for download.

**Comment by LunarWatcher** [comment]
No it can't. The original input has a changed format, and so does the output. #1 fixes the issues with output once it's fixed, which is going to be soon. The (programming) language doesn't matter

**Comment by TheOneWhoMade** [comment]
#5

**Comment by TheOneWhoMade** [comment]
I have an idea. Modify the save before it loads to include new vocab.

**Comment by LunarWatcher** [comment]
The vocab alone is USELESS. You can have a vocab to include every single word in the English (or any other language) and it'll not be enough. The network has to be trained to use them, meaning the issue is importing custom training data for the bot to train on. It'll add words if they're not found, assuming there's still space in the dict. The bot wouldn't be able to use the other words in the vocab because it wouldn't know *how to use them*.

**Comment by TheOneWhoMade** [comment]
Won't know how to use them as in it doesn't have experience with the word?

**Comment by TheOneWhoMade** [comment]
Is there any way to run a speed-training? As in no connection to the network + only in console?

**Comment by LunarWatcher** [comment]
The net doesn't need network for anything but installs. So yeah, you can disconnect from the network when you have the necessary packages

**Comment by TheOneWhoMade** [comment]
hmm. Combined with a grammar checker that could be powerful.

**Comment by LunarWatcher** [comment]
Grammar checking [exists](https://github.com/LunarWatcher/NN-chatbot/blob/master/grammarRules.py)

**Comment by TheOneWhoMade** [comment]
Yes I know it exists.

**Comment by TheOneWhoMade** [comment]
What I'm saying is run the bot output through the checker and if it is incorrect grammar then.. Oh you mean in the bot.

**Comment by TheOneWhoMade** [comment]
*cough* I'm impressed. You made a language parser with regex.

**Comment by TheOneWhoMade** [comment]
And these folks said it would be impossible: https://meta.stackoverflow.com/questions/362943/auto-capitalization-of-the-noun-i-on-post-submit

**Comment by TheOneWhoMade** [comment]
Or at least insanely difficult.

**Comment by LunarWatcher** [comment]
It is insanely difficult. My system will not get everything right. Nor will the [Magic Editor](https://github.com/SO-Close-Vote-Reviewers/UserScripts/blob/master/Magic%E2%84%A2Editor.user.js) (unrelated userscript for SE). But it works enough to get some level of sensibility in the output

**Comment by TheOneWhoMade** [comment]
Wow. Nobody likes me. (I WILL REMOVE THIS COMMENT)

**Comment by TheOneWhoMade** [comment]
So, could a method be used to set it to a random training and then remove that random training after the first epoch? So it would start out with  usage, and then get feedback, and then remove the fake usage and replace it with that feedback.

**Comment by TheOneWhoMade** [comment]
I tell you I'm pretty good with logic. (Even outside of Javascript)

**Comment by LunarWatcher** [comment]
There's no feedback on output atm, the closest you get is filtering future training data before teaching it

**Comment by TheOneWhoMade** [comment]
So how does it change and adapt?

**Comment by TheOneWhoMade** [comment]
How does it know whats right or wrong?

**Comment by TheOneWhoMade** [comment]
I'm sorry for incorrect terminology. I mean the results or the save that is produced by shutting down or an epoch

**Comment by LunarWatcher** [comment]
It doesn't train while it's being chatted with at the moment

**Comment by TheOneWhoMade** [comment]
So you have to train it in the console?

**Comment by TheOneWhoMade** [comment]
Well, I have yet to actually manage to install Tensorflow. It's actually quite difficult.

**Comment by LunarWatcher** [comment]
For now, yes. I'm trying to figure out continuous training, but it's not very well documented

**Comment by TheOneWhoMade** [comment]
... Feed me code for console usage.

**Comment by TheOneWhoMade** [comment]
... and then feed me code for chat usage.

**Comment by TheOneWhoMade** [comment]
All you should need to do is use the appropriate methods. 
`import os
`
all you need to do is use the os interface to log apropriate commands.

**Comment by TheOneWhoMade** [comment]
In fact, this could be used to make a `firstinstall.py` that would install all dependincies of the Chatbot. 😄

**Comment by TheOneWhoMade** [comment]
```
import os
os.system("Command you want to execute")
```
https://docs.python.org/2/library/os.html#os.system

**Comment by TheOneWhoMade** [comment]
?? Still here? ❔

**Comment by TheOneWhoMade** [comment]
So what you do is you make it so that the python not only chats but also trains at the same time.

**Comment by TheOneWhoMade** [comment]
https://github.com/LunarWatcher/NN-chatbot#known-issues

Is there any way to change the dimensions of a save?
